### LẤY DỮ LIỆU TỪ THỊ TRƯỜNG CHỨNG KHOÁN VIỆT NAM TỪ TCBS VÀ SSI THÔNG QUA MODULE VNSTOCK 

1. Gọi các module cần thiết

In [1]:
!pip install vnstock
!pip install openpyxl
import vnstock as vns
import pandas as pd
from IPython.display import display
pd.set_option('display.max_colwidth', 40)
pd.set_option('display.precision', 5)
pd.set_option('display.max_rows', 20)
pd.set_option('display.max_columns', 30)
import numpy as np
import openpyxl

Defaulting to user installation because normal site-packages is not writeable


2. Lấy danh sách các thông tin trên sàn chứng khoán Việt Nam

In [10]:
df1 = vns.listing_companies()
df1.to_csv("listing_companies.csv")
df1

ticker  group_code                                       company_name  \
0       VVS  UpcomIndex     Công ty Cổ phần Đầu tư Phát triển Máy Việt Nam   
1       XDC  UpcomIndex      Công ty TNHH MTV Xây dựng Công trình Tân Cảng   
2       HSV  UpcomIndex                   Công ty Cổ phần Gang Thép Hà Nội   
3       CST  UpcomIndex                 Công ty Cổ phần Than Cao Sơn - TKV   
4       BVL  UpcomIndex                            Công ty Cổ phần BV Land   
...     ...         ...                                                ...   
1625    XMP  UpcomIndex                Công ty Cổ phần Thủy điện Xuân Minh   
1626    YBC  UpcomIndex      Công ty Cổ phần Xi măng và Khoáng sản Yên Bái   
1627    YBM     VNINDEX     Công ty Cổ phần Khoáng sản Công nghiệp Yên Bái   
1628    YEG     VNINDEX                     Công ty Cổ phần Tập đoàn Yeah1   
1629    YTC  UpcomIndex  Công ty Cổ phần Xuất nhập khẩu Y tế Thành phố ...   

                  company_short_name  
0     Đầu tư Phát triển Máy Việt Nam  
1       Xây dựng Công trình Tân Cảng  
2                   Gang Thép Hà Nội  
3                 Than Cao Sơn - TKV  
4                            BV Land  
...                              ...  
1625             Thủy điện Xuân Minh  
1626   Xi măng và Khoáng sản Yên Bái  
1627           Khoáng sản CN Yên Bái  
1628                  Tập đoàn Yeah1  
1629                 XNK Y tế TP.HCM  

[1630 rows x 4 columns]

2.1 Xem tổng quát thông tin doanh nghiệp

In [11]:
TCB_data = vns.ticker_overview('TCB')
TCB_data

exchange    shortName  industryID industryIDv2   industry industryEn  \
0     HOSE  Techcombank         289         8355  Ngân hàng      Banks   

  establishedYear  noEmployees  noShareholders  foreignPercent  \
0            1993         9757            1901           0.225   

                         website  stockRating  deltaInWeek  deltaInMonth  \
0  http://www.techcombank.com.vn          4.2       -0.014        -0.021   

   deltaInYear  outstandingShare  issueShare companyType ticker  
0       -0.207            3517.2      3517.2          NH    TCB

2.2 Lấy dữ liệu lịch sử giá của cổ phiếu doanh nghiệp 

In [29]:
MCK="TCB"
def history_data(MCK):
    data = vns.stock_historical_data(symbol=MCK, 
                            start_date="2010-01-01", 
                            end_date='2023-02-25')
    return data

TCB_df = history_data(MCK)
TCB_df.to_csv("TCB_df_upto_2023_02_06.csv")
TCB_df.tail()

Open     High      Low    Close   Volume TradingDate
1166  29500.0  29500.0  28000.0  28200.0  5946569  2023-02-01
1167  28200.0  28400.0  27600.0  28000.0  3795867  2023-02-02
1168  28100.0  28200.0  27100.0  27100.0  6449574  2023-02-03
1169  27100.0  28000.0  27050.0  27950.0  3183964  2023-02-06
1170  27950.0  28050.0  27900.0  27900.0  1121700  2023-02-07

In [30]:
MCK="FPT"
FPT_df = history_data(MCK)
FPT_df.to_csv("FPT_df_upto_2023_02_06.csv")
FPT_df.tail()

Open     High      Low    Close   Volume TradingDate
3261  84000.0  84000.0  81800.0  82000.0  1051061  2023-02-01
3262  82100.0  82400.0  81300.0  81900.0  1327690  2023-02-02
3263  82000.0  82000.0  80100.0  80100.0  1730152  2023-02-03
3264  80000.0  81600.0  79700.0  80500.0   946115  2023-02-06
3265  80500.0  81900.0  80500.0  81000.0   231900  2023-02-07

2.3 Lấy một số chỉ báo kỹ thuật

In [39]:
vns.price_board('TCB,SSI,VND')

Mã CP  Giá Khớp Lệnh  KLBD/TB5D  T.độ GD  KLGD ròng(CM)  %KLGD ròng (CM)  \
0   TCB        27900.0       0.56     0.32              0              0.0   
1   SSI        19950.0       0.66     0.44        -121000            -21.8   
2   VND        15100.0       0.61     0.30         -82800            -15.5   

        RSI  MACD Volume MACD Signal  Tín hiệu KT Tín hiệu TB động     MA20  \
0  50.28044        -0.15        Sell         Sell      Strong Sell  28137.5   
1  52.06900        -0.08        Sell  Strong Sell      Strong Sell  19900.0   
2  50.36031        -0.12        Sell         Sell      Strong Sell  15257.5   

      MA50    MA100  Phiên +/-   ...  Đáy 1Y  %Đỉnh 1Y  %Đáy 1Y  \
0  27267.0  27339.5           1  ...   20700     -48.0     35.0   
1  19255.0  18340.5           1  ...   13900     -52.9     43.9   
2  14445.0  14319.2           1  ...    9720     -56.4     55.9   

   %Giá - %VNI (1M)  %Giá - %VNI (1Y)  Khối lượng Dư mua  Khối lượng Dư bán  \
0              -2.1             -20.7             184700             246500   
1               0.5             -24.5              81500             162500   
2               0.6             -18.4             625300             287400   

   Khớp nhiều nhất         seq  vnid3d  vnid1m  vnid3m  vnid1y    vnipe  \
0            27950  1675744140     1.1     3.2     9.2   -26.3  12.2986   
1            20200  1675744140     1.1     3.2     9.2   -26.3  12.2986   
2            15300  1675744140     1.1     3.2     9.2   -26.3  12.2986   

     vnipb  
0  1.62374  
1  1.62374  
2  1.62374  

[3 rows x 50 columns]

2.4 Lấy dữ liệu giao dịch theo thời gian thực

In [43]:
df =  vns.stock_intraday_data(symbol='TCB', 
                            page_num=0, 
                            page_size=5000)
print(df.head(10))

     price  volume    cp  rcp   a        ba        sa     hl  pcp      time
0  27900.0     200 -50.0  0.0  SD  502500.0  583900.0  False  0.0  11:29:59
1  27900.0    9000 -50.0  0.0  SD  502500.0  603700.0   True  0.0  11:29:59
2  27900.0     700 -50.0  0.0  SD  502500.0  594600.0  False  0.0  11:29:59
3  27900.0     100 -50.0  0.0  SD  502500.0  594700.0  False  0.0  11:29:59
4  27900.0   10000 -50.0  0.0  SD  502500.0  593900.0   True  0.0  11:29:59
5  27900.0     500 -50.0  0.0  SD  502500.0  583700.0  False  0.0  11:29:40
6  27900.0    1000 -50.0  0.0  SD  502500.0  583200.0  False  0.0  11:29:38
7  27900.0    6000 -50.0  0.0  SD  502500.0  582200.0  False  0.0  11:29:28
8  27900.0    2300 -50.0  0.0  SD  502500.0  576200.0  False  0.0  11:29:26
9  27900.0    3700 -50.0  0.0  SD  502500.0  573900.0  False  0.0  11:29:26


2.5 So sánh chỉ số cơ bản của các doanh nghiệp

In [20]:
a = vns.financial_ratio_compare (symbol_ls=["MBB", "TCB"], industry_comparison='true', frequency= 'Yearly', start_year=2010)
a.head(20)

Chỉ số         2017         2018         2019         2020  \
0          P/E          NaN          NaN          NaN          NaN   
1          TCB     15.89460     10.99041      7.71236     11.10489   
2          MBB     17.02217      7.83569      6.33920      9.36571   
3          TCB     15.89460     10.99041      7.71236     11.10489   
4   Toàn ngành     20.32825     12.21110      8.02101     13.62224   
5          P/B          NaN          NaN          NaN          NaN   
6          TCB      3.80421      1.79855      1.25773      1.84628   
7          MBB      2.12681      1.48583      1.30508      1.61535   
8          TCB      3.80421      1.79855      1.25773      1.84628   
9   Toàn ngành      3.19040      1.85043      1.28993      2.08961   
10   EPS (VND)          NaN          NaN          NaN          NaN   
11         TCB   5530.18037   2420.29225   2878.49600   3516.46729   
12         MBB   1938.64860   2865.09616   3296.94715   2952.26034   
13         TCB   5530.18037   2420.29225   2878.49600   3516.46729   
14  Toàn ngành   2371.07368   2465.84643   2968.27496   2776.74433   
15  BVPS (VND)          NaN          NaN          NaN          NaN   
16         TCB  23105.99369  14789.65377  17650.89282  21150.61367   
17         MBB  15516.19345  15109.41371  16014.38519  17117.06508   
18         TCB  23105.99369  14789.65377  17650.89282  21150.61367   
19  Toàn ngành  14980.53646  15156.36817  17317.87325  18260.44694   

           2021         2022  
0           NaN          NaN  
1       9.79056      4.87866  
2       9.79019      4.84967  
3       9.79056      4.87866  
4      13.42349      8.34193  
5           NaN          NaN  
6       1.91516      0.87543  
7       2.08657      1.11635  
8       1.91516      0.87543  
9       2.24843      1.53360  
10          NaN          NaN  
11   5137.60260   5729.03344  
12   3360.50651   3855.93044  
13   5137.60260   5729.03344  
14   2851.85183   3398.12209  
15          NaN          NaN  
16  26264.09761  31927.33475  
17  15767.53865  16751.06459  
18  26264.09761  31927.33475  
19  17279.31947  18813.76808

In [25]:
vns.valuation_rating("HPG")

industryEn ticker  valuation  pe  pb  ps  evebitda  dividendRate
0      Steel    HPG        2.4   2   3   3         2             2

In [27]:
vns.market_top_mover('ForeignTrading')

foreignBuyVolume  foreignBuyValue  foreignSellVolume  foreignSellValue  \
0          2811800.0      7.35515e+10           188500.0       4.92034e+09   
1           686800.0      2.06032e+10           137400.0       4.11570e+09   
2           506500.0      5.37458e+09                0.0       0.00000e+00   
3           326700.0      1.29261e+10             6200.0       2.45250e+08   
4           265300.0      6.33110e+09             5000.0       1.19200e+08   
..               ...              ...                ...               ...   
25           57400.0      6.15215e+09             3150.0       3.36343e+08   
26           53200.0      3.84139e+08                0.0       0.00000e+00   
27           57500.0      1.09292e+09             5700.0       1.07950e+08   
28           93800.0      4.06156e+09            43900.0       1.89820e+09   
29           48400.0      9.22960e+08                0.0       0.00000e+00   

   organCode ticker                               sectorName        value  \
0        STB    STB                                Ngân hàng  1.87292e+11   
1        CTG    CTG                                Ngân hàng  4.58740e+10   
2        LCG    LCG                                 Xây dựng  2.97330e+10   
3      PETRO    PLX            Sản xuất và Khai thác dầu khí  3.32070e+10   
4        PVS    PVS              Thiết bị và Dịch vụ Dầu khí  5.12370e+10   
..       ...    ...                                      ...          ...   
25       GAS    GAS             Phân phối xăng dầu & khí đốt  9.40700e+09   
26     DXRES    DXS  Tư Vấn, Định giá, Môi giới Bất động sản  8.01900e+09   
27       OCB    OCB                                Ngân hàng  7.44200e+09   
28       DPM    DPM  Sản phẩm hóa dầu, Nông dược & Hóa ch...  1.71830e+10   
29     BAFVN    BAF                Nuôi trồng nông & hải sản  1.24080e+10   

       volume     price  floorPrice  ceilingPrice  referencePrice         o  \
0   7170000.0   26000.0     23950.0       27450.0         25700.0   25900.0   
1   1528800.0   29950.0     27850.0       31950.0         29900.0   29800.0   
2   2800500.0   10500.0     10000.0       11500.0         10750.0   10750.0   
3    839200.0   39500.0     36100.0       41500.0         38800.0   39000.0   
4   2146000.0   23800.0     21200.0       25800.0         23500.0   23200.0   
..        ...       ...         ...           ...             ...       ...   
25    87800.0  107000.0     98600.0      113400.0        106000.0  106700.0   
26  1105500.0    7260.0      6570.0        7550.0          7060.0    7090.0   
27   391900.0   18850.0     18050.0       20750.0         19400.0   18850.0   
28   397000.0   43200.0     40100.0       46100.0         43100.0   43300.0   
29   651200.0   19000.0     17850.0       20450.0         19150.0   19150.0   

           h         l  priceChange  percentPriceChange  rank  totalRank  \
0    26300.0   25750.0        300.0              0.0117     4         27   
1    30200.0   29600.0         50.0              0.0017     1         27   
2    10800.0   10500.0       -250.0             -0.0233    47        354   
3    40000.0   39000.0        700.0              0.0180     2          3   
4    24100.0   23100.0        300.0              0.0128     1          9   
..       ...       ...          ...                 ...   ...        ...   
25  107800.0  105000.0       1000.0              0.0094     1        101   
26    7390.0    7040.0        200.0              0.0283    65        131   
27   19200.0   18850.0       -550.0             -0.0284    17         27   
28   43450.0   43150.0        100.0              0.0023     3         74   
29   19200.0   18900.0       -150.0             -0.0078    16        112   

                                performance  \
0   {'organCode': 'STB', 'percentPriceCh...   
1   {'organCode': 'CTG', 'percentPriceCh...   
2   {'organCode': 'LCG', 'percentPriceCh...   
3   {'organCode': 'PETRO', 'percentPrice...   
4   {'organCode': 'PVS', 'percentPriceCh...  

2.10 Mức độ ảnh hưởng đến chỉ số chung

In [28]:
vns.fr_trade_heatmap('All', 'FrBuyVol')

organCode  name      value  percentPriceChange  referencePrice  matchPrice  \
0      PETRO   PLX   326700.0             0.01800         38800.0     39500.0   
1        PVS   PVS   265300.0             0.01280         23500.0     23800.0   
2        PVD   PVD    16200.0             0.00260         19350.0     19400.0   
3       BSRC   BSR      200.0             0.02100         15900.0     16200.0   
0        HPG   HPG  1117700.0            -0.02840         21150.0     20550.0   
..       ...   ...        ...                 ...             ...         ...   
9        SHB   SHB     4300.0            -0.00480         10450.0     10400.0   
10      None  Khác     4100.0            -0.00239             0.0         0.0   
0        FPT   FPT    63900.0             0.00620         80500.0     81000.0   
1        SAM   SAM     6400.0             0.00000          6060.0      6060.0   
2        ELC   ELC      300.0            -0.01200         12500.0     12350.0   

    ceilingPrice  floorPrice        industry_name     rate  
0        41500.0     36100.0              Dầu khí  0.01614  
1        25800.0     21200.0              Dầu khí  0.01614  
2        20700.0     18000.0              Dầu khí  0.01614  
3        18200.0     13600.0              Dầu khí  0.01614  
0        22600.0     19700.0      Nguyên vật liệu -0.01333  
..           ...         ...                  ...      ...  
9        11150.0      9720.0            Ngân hàng -0.00480  
10           0.0         0.0            Ngân hàng -0.00480  
0        86100.0     74900.0  Công nghệ Thông tin  0.00551  
1         6480.0      5640.0  Công nghệ Thông tin  0.00551  
2        13350.0     11650.0  Công nghệ Thông tin  0.00551  

[89 rows x 10 columns]